llama.cpp

```diff
@@ -4329,6 +4351,32 @@ static void llm_load_print_meta(llama_model_loader & ml, llama_model & model) {
     if (vocab.linefeed_id    != -1) { LLAMA_LOG_INFO( "%s: LF token         = %d '%s'\n", __func__, vocab.linefeed_id,    vocab.id_to_token[vocab.linefeed_id].text.c_str() );    }
 }
 
+static int llm_determine_max_ngl(const llama_model_loader & ml, const llama_model & model, const int main_gpu) {
+    const auto & hparams = model.hparams;
+
+    size_t available_gpu_memory = llama_get_available_device_memory(main_gpu);
+
+    // TODO: This is a rough, pretty inaccurate estimate, should implement using existing layer size and not guesstimating
+    size_t model_size = ml.n_bytes;
+    int32_t model_layers = hparams.n_layer;
+    size_t memory_per_layer = model_size / model_layers;
+
+    // TODO: get buffer size dynamically
+    int32_t buf_size = 400 * MiB;
+    int32_t buf_size_k = 200 * MiB;
+    int32_t buf_size_v = 200 * MiB;
+
+    int32_t total_buf_size = buf_size + buf_size_k + buf_size_v;
+
+    available_gpu_memory = available_gpu_memory - hparams.n_ctx_train; // context size
+    available_gpu_memory = available_gpu_memory - total_buf_size; // buffer size
+
+    // Calculate the maximum number of layers that can fit into the GPU memory
+    int32_t max_ngl = std::floor(static_cast<float>(available_gpu_memory) / memory_per_layer);
```

Comment:
You should leave a small amount of headroom when the number of layers is determined automatically; you would want to avoid a scenario where an application ooms because llama.cpp only left 5 MB of VRAM.

Reply:
I thought about doing that; thanks for reminding